### PR TITLE
EOS-21324: copy-object causes s3server crash

### PR DESCRIPTION
**Pull request by evgeniy-brazhnikov**
Signed-off-by: evgeniy-brazhnikov <evgeniy.brazhnikov@seagate.com>

### CODE REVIEW

File: server/s3_object_data_copier.cc [hunk 1]
```diff
@@ -189,19 +187,20 @@ void S3ObjectDataCopier::write_data_block() {
...
 
   assert(!write_in_progress);
-  assert(data_blocks_writing.empty());
+
   assert(!data_blocks_read.empty());
 
-  data_blocks_writing = std::move(data_blocks_read);
+  assert(p_evbuffer_read != nullptr);
+  assert(!p_evbuffer_write);
+  p_evbuffer_write = p_evbuffer_read;
```
Comment (knrajnambiar76) [inline server/s3_object_data_copier.cc:195]: If I get this correct, you are passing evbuffers of read to write

File: server/s3_object_data_copier.h [hunk 2]
```diff
@@ -63,9 +62,9 @@ class S3ObjectDataCopier {
   bool write_in_progress;
   // Size of each ev buffer (e.g, 16384)
   size_t size_of_ev_buffer;
-  // Queue of ev Buffers containing read data blocks,
-  // which will be freed after write data completion
-  std::deque<struct evbuffer*> read_data_buffer;
+
+  struct evbuffer* p_evbuffer_read = nullptr;
```
Comment (knrajnambiar76) [inline server/s3_object_data_copier.h:66]: You can put coment about how p_evbuffer_read and p_evbuffer_write are related

**Review by knrajnambiar76 — COMMENTED** [review]
Once test with uploading a zero byte file and then copying the zero byte file and downloading the same

File: server/s3_object_data_copier.cc [hunk 3]
```diff
@@ -218,27 +217,15 @@ void S3ObjectDataCopier::write_data_block() {
...
-}
-
 void S3ObjectDataCopier::write_data_block_success() {
   s3_log(S3_LOG_INFO, request_id, "%s Entry\n", __func__);
 
   assert(write_in_progress);
   write_in_progress = false;
 
-  cleanup_blocks_written();
+  assert(p_evbuffer_write != nullptr);
```
Comment (knrajnambiar76) [inline server/s3_object_data_copier.cc:226]: Once test with copying 0 byte file

File: server/s3_object_data_copier.cc (hunk 1 above)
Comment (evgeniy-brazhnikov) [reply server/s3_object_data_copier.cc:195]: Not exactly. These variables are used only to hold evbuffer's and to free them in timely manner.

File: server/s3_object_data_copier.h (hunk 2 above)
Comment (evgeniy-brazhnikov) [reply server/s3_object_data_copier.h:66]: p_evbuffer_write holds blocks of data which are currently being written.
p_evbuffer_read contains blocks of data which has already been read and ready to write.

File: server/s3_object_data_copier.cc (hunk 3 above)
Comment (evgeniy-brazhnikov) [reply server/s3_object_data_copier.cc:226]: Succeed

**Comment by DPG17** [comment]
Evgeniy Brazhnikov: A general comment: why the crash happened? which code line resulted in crash?

**Review by knrajnambiar76 — APPROVED** [review]

**Review by DPG17 — APPROVED** [review]

**Comment by nileshgovande** [comment]
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Component-Test/job/S3Server-PreMerge-Test/1589/
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Mini-Provisioning/job/cortx-s3server/1381/

**Pull request merged**